### PR TITLE
More compatibility tweaks for pyro-api

### PR DESCRIPTION
**Pull request by neerajprad**
Follow up on https://github.com/pyro-ppl/pyro-api/pull/8 in pyro-api, this increases the coverage of tests with the following changes to the compatibility wrapper:

 - Added support for `ClippedAdam` optimizer.
 - Wrappers to allow for torch style zeros/ones array initializers. i.e. in pytorch, `torch.ones(2, 3)`, `torch.ones((2, 3))` and `torch.ones([2, 3])` can all be used to initialize a tensor of size (2, 3). In NumPy, the first invocation isn't allowed.
 - Allow accessing SVI parameters via the global param store when `pyro.param` is used outside an inference context. This doesn't support the full param store API e.g. initializing using constraints etc., which are handled by inference algorithms directly in NumPyro. This simply updates the global param store when `svi.step` is called. 

With these fixes, all tests pass expect `test_mean_field ` which isn't implemented.

### CODE REVIEW

File: numpyro/compat/pyro.py [hunk 1]
```diff
@@ -1,13 +1,33 @@
...
+    return _PARAM_STORE
+
+
+def clear_param_store():
+    return _PARAM_STORE.clear()
+
+
+def param(name, *args, **kwargs):
+    # Get value assuming statement is wrapped in a substitute handler.
+    val = _param(name, *args, **kwargs)
```
Comment (fehiepsi) [inline numpyro/compat/pyro.py]: Look reasonable to me!

File: numpyro/compat/infer.py [hunk 2]
```diff
@@ -131,6 +132,8 @@ def step(self, *args, rng=None, **kwargs):
                                 'dicts of arrays.')
             else:
                 raise e
+        params = super(SVI, self).get_params(self.svi_state)
```
Comment (fehiepsi) [inline numpyro/compat/infer.py]: I think we should `jit` this.

File: numpyro/compat/ops.py [hunk 3]
```diff
@@ -1,3 +1,34 @@
...
+
+def ones(*sizes, **kwargs):
+    if len(sizes) == 0:
+        raise ValueError('Positional `size` argument not provided.')
+    elif len(sizes) == 1:
+        if isinstance(sizes[0], (tuple, list)):
+            sizes = sizes[0]
+    if not all([isinstance(s, int) for s in sizes]):
+        raise ValueError('Invalid data type for `size` provided.')
+    return np.ones(sizes, **kwargs)
```
Comment (fehiepsi) [inline numpyro/compat/ops.py]: lgtm

File: numpyro/optim.py [hunk 4]
```diff
@@ -75,6 +77,27 @@ def __init__(self, *args, **kwargs):
...
+class ClippedAdam(_NumpyroOptim):
+    """
+    :class:`~numpyro.optim.Adam` optimizer with gradient clipping.
+
+    **Reference:**
+
+    `A Method for Stochastic Optimization`, Diederik P. Kingma, Jimmy Ba
+    https://arxiv.org/abs/1412.6980
+    """
+    def __init__(self, *args, **kwargs):
```
Comment (fehiepsi) [inline numpyro/optim.py]: I think we should expose `clip_norm` here and add docs for it.

File: numpyro/compat/ops.py [hunk 5]
```diff
@@ -1,3 +1,34 @@
...
+
+import jax.numpy as np
 from jax.numpy import *  # noqa: F401, F403
 
 tensor = array  # noqa: F405
+
+randn = onp.random.randn
+
+
+# Provide wrappers to initialize ones/zeros using the pytorch convention
```
Comment (fritzo) [inline numpyro/compat/ops.py]: I think it's fine to restrict to numpy-style usage in pyro-api, but if you want to maintain these that's fine too. Let me know either way, and we can switch the pyroapi.tests to use numpy-style invocations.

File: numpyro/optim.py [hunk 6]
```diff
@@ -75,6 +77,27 @@ def __init__(self, *args, **kwargs):
         super(Adam, self).__init__(optimizers.adam, *args, **kwargs)
 
 
+class ClippedAdam(_NumpyroOptim):
```
Comment (fritzo) [inline numpyro/optim.py]: Note Martin's `ClippedAdam` also includes learning rate decay. We don't need to implement that now, but can you make sure this errors if an `lrd` kwarg is found?

Comment (neerajprad) [reply numpyro/optim.py]: I think we can support lrd too, because we can provide a schedule for the learning rate to the jax optimizer. I will try to include that or error if it turns out to be more complicated.

File: numpyro/compat/infer.py (hunk 2 above)
Comment (neerajprad) [reply numpyro/compat/infer.py]: Sure, will do that. Do you think this will have significant overhead?

File: numpyro/compat/ops.py (hunk 5 above)
Comment (neerajprad) [reply numpyro/compat/ops.py]: I don't mind maintaining these wrappers, my hope is that this will be relatively hassle free, lets see. :smile: There is a bit of an inconsistency in the numpy API too, because `torch.random.randn()` doesn't accept a tuple size arg, so this provides a more consistent interface.

File: numpyro/compat/infer.py (hunk 2 above)
Comment (fehiepsi) [reply numpyro/compat/infer.py]: Yes, especially for models with many constrained supports.

File: numpyro/optim.py (hunk 4 above)
Comment (neerajprad) [reply numpyro/optim.py]: done.

File: numpyro/optim.py (hunk 6 above)
Comment (neerajprad) [reply numpyro/optim.py]: Done.

File: numpyro/compat/infer.py (hunk 2 above)
Comment (neerajprad) [reply numpyro/compat/infer.py]: Yeah, that's a good point. I wish there was a way to signal the end of learning with Pyro's interface. Pyro's param store will do this lazily, but we'll be inefficiently doing this per `svi.step`. :(

Comment (fritzo) [reply numpyro/compat/infer.py]: Can you explain for me how this is less efficient than Pyro? Even in Pyro constrained params must be transformed once per svi step.

Comment (fehiepsi) [reply numpyro/compat/infer.py]: I meant this will have significant overhead when we didn't jit it. In NumPyro, params are already transformed to compute elbo. The transform here is not necessary unless we want to access param values from the param store. But I think this overhead is small because we already jit it.

Comment (neerajprad) [reply numpyro/compat/infer.py]: Actually, I think the only difference is that pyro will compute and cache this when it calls `pyro.param` whereas we will be doing this at least twice (once in inference and then again to put the parameters in the param store). But hopefully, with JIT, this shouldn't be as much of an issue.

Comment (fehiepsi) [reply numpyro/compat/infer.py]: @neerajprad How about we resolve this by changing [this line](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/svi.py#L205) to
```
def loss_fn(x):
     constrained_params = self.constrain_fn(x)
     return self.loss.loss(rng_seed, constrained_params, ...), constrained_params

(loss, constrained_params), grad = value_and_grad(loss_fn, has_aux=True)(params)
```
Basically, we use `has_aux=True` for constrained_params and move it to `svi_state`.

Comment (neerajprad) [reply numpyro/compat/infer.py]: That sounds reasonable. You mean, add another field called `constrained_params` to `SVIState`?

**Comment by neerajprad** [comment]
@fehiepsi - I will address your suggestion regarding the optimization in a separate PR, if that's okay?
